### ISSUE TITLE
Release 1.10

### ISSUE THREAD

**Issue by ysitu** [networkx/networkx#1590]
@hagberg Let's release 1.10 so that we can unblock PRs from @MridulS. I have set up [1.10rc1](https://github.com/networkx/networkx/releases/tag/networkx-1.10rc1) tag and prerelease.

**Comment by hagberg** [comment]
Yes, fine.  There are documentation format issues that are breaking the automatic nightly builds #1571.
It would be nice to clean those before we release 1.10.

**Comment by chebee7i** [comment]
What is the standard procedure here? There are two PRs that we want to cherry-pick into the 1.10 release. I can do that, but should I create a new branch RC2 and put them there, or should I add them to the RC1 branch?

**Comment by hagberg** [comment]
I don't have any strong opinions other than to keep it simple.  In the past the pace of development was slow enough that we could make some RC versions on the master branch and not have a lot of extra additions in between.  Maybe we need to make a networkx-1.10 branch and put tags on that for RCs?

**Comment by chebee7i** [comment]
Yeah, I think that makes sense. I think tags are independent of the branches, since they are tied to a revision. So I will checkout the tag for 1.10rc1, then push up a branch `networkx-1.0`. From there, I'll cherry-pick the desired commits. At some point, we can tag for 1.10rc2 or whatever.

**Comment by chebee7i** [comment]
I think we are good for the 1.10 release. We are going to have to keep this branch alive for a while, b/c I suspect we'll have to keep support for the 1.x branch for a while (and maybe even put out bugfixes for it, but hopefully not).

**Comment by hagberg** [comment]
Shall we make make an rc2 and ask people to test?  Maybe we could get @santrotosi to take a look to see if we need to fix anything for Debian packaging?  He has been very helpful and I'd like to make it as easy as possible for him

**Comment by ysitu** [comment]
What is the current status?

**Comment by hagberg** [comment]
It looks like we are close.  How about we make an rc2 and ask people to test?  There are often some small issues that can be easily resolved regarding packaging or documentation.

**Comment by hagberg** [comment]
OK everyone - please test.  I added a Python wheel there too.
https://github.com/networkx/networkx/releases/tag/networkx-1.10rc2
It might be good to fix the release.py file so it reflects 1.10rc2 instead of 2.0dev (I did that in the wheel).

**Comment by hagberg** [comment]
I pushed the rc2 version to pypi https://pypi.python.org/pypi/networkx/
Would someone send a note to the networkx-discuss mailing list and encourage people to try it out?

**Comment by ysitu** [comment]
The version number is now fixed in the branch.

**Comment by hagberg** [comment]
No comments from the community.  I suggest we tag and release.

**Comment by ysitu** [comment]
Add one final commit to enable Himanshu's work?

**Comment by chebee7i** [comment]
I'd be fine with that. If we ultimately decide against namespaces, we can always remove `networkx.addons` in 2.0 or later. The perturbation required to networkx seems fairly minimal.

**Comment by OrkoHunter** [comment]
We also had discussions to separate the `drawing` package and publish it as a separate `addon` to give it a better exposure for development. I remember there was a good support for #1325 . So, now or later, we should have `networkx.addons` for the plotting module because that would not be properly an *external* package for NetworkX core.

**Comment by ysitu** [comment]
@OrkoHunter Do you already have a PR that adds `networkx/addons/__init__.py`?

**Comment by OrkoHunter** [comment]
That PR got merged already

**Comment by ysitu** [comment]
@OrkoHunter I mean for the 1.10 branch.

**Comment by OrkoHunter** [comment]
No. I thought we would take master branch and make 1.10rc3. Should I send a PR to the `1.10` branch?

**Comment by jtorrents** [comment]
I agree that we should go ahead and enable `networkx.addons` for 1.10. Maybe we could say in the release notes that this is an experimental feature and might be subject to change in the future.

**Comment by ysitu** [comment]
@OrkoHunter You can cherrypick the commit from master to 1.10.

**Comment by OrkoHunter** [comment]
```sh
$ git cherry-pick bdea25fc5a988e8c99382493b27bfb8f811ea8d1                                                                                     [v1.10] 22:07:51
error: Commit bdea25fc5a988e8c99382493b27bfb8f811ea8d1 is a merge but no -m option was given.
fatal: cherry-pick failed
```
Should I just add a new commit?

**Comment by ysitu** [comment]
You should cherrypick the commit that truly added the file, not the merge commit.

**Comment by OrkoHunter** [comment]
#1646

**Comment by hagberg** [comment]
I still think the addons namespacery isn't clearly a good idea.  Can anyone provide some examples of projects that are doing this happily and what value it adds for the extra complexity?

**Comment by jfinkels** [comment]
Not sure whether it's good or bad, but here's how [Flask][1] does it: http://flask.pocoo.org/docs/0.10/extensiondev/#extension-import-transition

Essentially, when "addons" are created, they live in a Python package or module named `flask_xxx`, then when they are imported by the end user, they are imported as `import flask.ext.xxx`.

[1]: https://github.com/mitsuhiko/flask

**Comment by hagberg** [comment]
OK - that is a different approach that doesn't use namespaces.   It uses this trickery instead https://github.com/mitsuhiko/flask/blob/master/flask/ext/__init__.py

**Comment by OrkoHunter** [comment]
Flask switched on to this "new" method from namespace package systems (similar to which, @chebee7i and me were talking about few weeks ago for `networkx-metis`. But later on we decided that namespace package systems is kind of more suited for this). 

In `flask/ext/__init__.py` it says,
> We're switching from namespace packages because it was just too painful for everybody involved.

I can't understand how this can be painful and the ExtensionImporter system not. I see both as similar method instead the first one uses the built-in facility provided by `import`.

Anyways, both cases will work and will require `networkx/addons/__init__.py`. I'd be okay switching to it. But still, this is coming to me as a choice which does not have a good reason behind it.

**Comment by chebee7i** [comment]
@OrkoHunter but there are obvious weird parts about namespaces that don't seem to have a great support. Suppose, I install networkx as `pip install -e .` Then, suppose I install networkx-metis as `pip install .` and networkx-lemon as `pip install -e .`  What happens in these cases? Could we even uninstall properly? In one of the other conversations, I linked to an issue that seemed to indicate this was problematic for namespaces. If we went with an external package, then there would be no issues.

When it comes down to it, what really is gained by using a namespace? Is it just that the addon looks as if it is more "official" since it is in `networkx.addons`?

**Comment by OrkoHunter** [comment]
I read the issue now, about using namespace for editable installation https://github.com/pypa/pip/issues/3. Seems like it's really a considerable problem. Okay so let's switch.

**Comment by hagberg** [comment]
I'm ready to release this if someone wants to make a tag.  No issues reported with rc2.

**Comment by jtorrents** [comment]
@hagberg in #1660 I cherry-picked the two commits that fixed the sporadic test failures on Travis for k-components approximation, and tagged the branch `v1.10` as release (by setting `dev=False` in `release.py`). 

I think that before releasing we still need to decide how to approach the `addons` feature in v1.10. Is that right @ysitu @OrkoHunter ? I didn't follow closely the discussion on that.

**Comment by OrkoHunter** [comment]
@jtorrents We can make a release without `networkx/addons`. As of now, we decided to import the addons from its own package without using the `networkx.addons` namespace.

**Comment by jtorrents** [comment]
@OrkoHunter , sounds good. Then I think we are ready to release 1.10 once #1660 is merged into the `v1.10` branch.

**Comment by hagberg** [comment]
#1660 is merged.  Anything else?  I'll make a the 1.10 release later today.

**Comment by harlowja** [comment]
How goes this? I see a rc2 release on pypi still, is a non-rc release expected?

**Comment by hagberg** [comment]
I think it is just waiting for me.  I'll do this ASAP.

**Comment by hagberg** [comment]
I just made the release.

**Comment by harlowja** [comment]
Thanks, updated in openstack @ https://review.openstack.org/208220 so that openstack libraries/projects can use it, much appreciated.
